File: examples/test_mcp_connection.js
#!/usr/bin/env node
/**
 * Simple script to test MCP connection to APM using Node.js
 * 
 * Usage:
 * 1. Start APM with MCP enabled:
 *    APM_MCP_ENABLED=1 cargo run -- start
 * 
 * 2. Run this script:
 *    node examples/test_mcp_connection.js
 */

const net = require('net');

class MCPClient {
    constructor(host = '127.0.0.1', port = 7338) {
        this.host = host;
        this.port = port;
        this.socket = null;
        this.buffer = '';
        this.messageId = 0;
        this.pendingCallbacks = new Map();
    }

    connect() {
        return new Promise((resolve, reject) => {
            this.socket = net.createConnection({ host: this.host, port: this.port }, () => {
                console.log(`Connected to MCP server at ${this.host}:${this.port}`);
                resolve();
            });

            this.socket.on('data', (data) => {
                this.buffer += data.toString();
                this.processBuffer();
            });

            this.socket.on('error', reject);
            this.socket.on('close', () => {
                console.log('Connection closed');
            });
        });
    }

    processBuffer() {
        const lines = this.buffer.split('\n');
        this.buffer = lines.pop() || '';

        for (const line of lines) {
            if (line.trim()) {
                try {
                    const response = JSON.parse(line);
                    const callback = this.pendingCallbacks.get(response.id);
                    if (callback) {
                        this.pendingCallbacks.delete(response.id);
                        callback(response);
                    }
                } catch (e) {
                    console.error('Failed to parse response:', e);
                }
            }
        }
    }

    sendRequest(method, params = {}) {
        return new Promise((resolve) => {
            const id = ++this.messageId;
            const request = {
                jsonrpc: '2.0',
                method,
                params,
                id
            };

            this.pendingCallbacks.set(id, resolve);
            this.socket.write(JSON.stringify(request) + '\n');
        });
    }

    close() {
        if (this.socket) {
            this.socket.end();
        }
    }
}

async function testMCPConnection() {
    const client = new MCPClient();

    try {
        // Connect to server
        await client.connect();

        // 1. Initialize
        console.log('\n1. Initializing MCP connection...');
        const initResponse = await client.sendRequest('initialize', {
            protocolVersion: '0.1.0',
            capabilities: {},
            clientInfo: {
                name: 'node-test-client',
                version: '1.0.0'
            }
        });
        console.log('Response:', JSON.stringify(initResponse, null, 2));

        // Send initialized notification (required by MCP protocol)
        const initializedMsg = {
            jsonrpc: '2.0',
            method: 'notifications/initialized',
            params: {}
        };
        client.socket.write(JSON.stringify(initializedMsg) + '\n');
        console.log('\nSent initialized notification');

        // 2. List tools
        console.log('\n2. Listing available tools...');
        const toolsResponse = await client.sendRequest('tools/list', {});
        console.log('Response:', JSON.stringify(toolsResponse, null, 2));

        // 3. Spawn a test process
        console.log('\n3. Spawning a test process...');
        const spawnResponse = await client.sendRequest('tools/call', {
            name: 'spawn',
            arguments: {
                name: 'test-node-echo',
                command: 'echo',
                args: ['Hello from Node.js MCP client!']
            }
        });
        console.log('Response:', JSON.stringify(spawnResponse, null, 2));

        // 4. List processes
        console.log('\n4. Listing processes...');
        const listResponse = await client.sendRequest('tools/call', {
            name: 'list',
            arguments: {}
        });
        console.log('Response:', JSON.stringify(listResponse, null, 2));

        // 5. Query system overview
        console.log('\n5. Querying system overview...');
        const queryResponse = await client.sendRequest('tools/call', {
            name: 'query',
            arguments: {
                type: 'system_overview'
            }
        });
        console.log('Response:', JSON.stringify(queryResponse, null, 2));

        console.log('\nMCP connection test completed successfully!');

    } catch (error) {
        console.error('ERROR:', error.message);
        if (error.code === 'ECONNREFUSED') {
            console.error('Make sure APM is running with MCP enabled:');
            console.error('  APM_MCP_ENABLED=1 cargo run -- start');
        }
        process.exit(1);
    } finally {
        client.close();
    }
}

// Run the test
testMCPConnection();

File: examples/test_mcp_connection.py
#!/usr/bin/env python3
"""
Simple script to test MCP connection to APM.

This demonstrates basic MCP protocol communication with APM's TCP server.

Usage:
1. Start APM with MCP enabled:
   APM_MCP_ENABLED=1 cargo run -- start

2. Run this script:
   python examples/test_mcp_connection.py
"""

import json
import socket
import sys

def send_json_rpc(sock, method, params, msg_id=1):
    """Send a JSON-RPC message and receive response."""
    message = {
        "jsonrpc": "2.0",
        "method": method,
        "params": params,
        "id": msg_id
    }
    
    # Send message
    data = json.dumps(message) + "\n"
    sock.sendall(data.encode())
    
    # Receive response
    response = ""
    while True:
        chunk = sock.recv(1024).decode()
        response += chunk
        if "\n" in response:
            break
    
    # Parse first complete JSON response
    lines = response.split("\n")
    if lines[0]:
        return json.loads(lines[0])
    return None

def test_mcp_connection():
    """Test basic MCP operations."""
    # Connect to APM's MCP server
    HOST = '127.0.0.1'
    PORT = 7338  # Default MCP port
    
    print(f"Connecting to MCP server at {HOST}:{PORT}...")
    
    try:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
            sock.connect((HOST, PORT))
            print("Connected!")
            
            # 1. Initialize
            print("\n1. Initializing MCP connection...")
            response = send_json_rpc(sock, "initialize", {
                "protocolVersion": "0.1.0",
                "capabilities": {},
                "clientInfo": {
                    "name": "test-client",
                    "version": "1.0.0"
                }
            }, msg_id=1)
            print(f"Response: {json.dumps(response, indent=2)}")
            
            # Send initialized notification (required by MCP protocol)
            initialized_msg = {
                "jsonrpc": "2.0",
                "method": "notifications/initialized",
                "params": {}
            }
            sock.sendall((json.dumps(initialized_msg) + "\n").encode())
            print("\nSent initialized notification")
            
            # 2. List tools
            print("\n2. Listing available tools...")
            response = send_json_rpc(sock, "tools/list", {}, msg_id=2)
            print(f"Response: {json.dumps(response, indent=2)}")
            
            # 3. Spawn a test process
            print("\n3. Spawning a test process...")
            response = send_json_rpc(sock, "tools/call", {
                "name": "spawn",
                "arguments": {
                    "name": "test-echo",
                    "command": "echo",
                    "args": ["Hello from Python MCP client!"]
                }
            }, msg_id=3)
            print(f"Response: {json.dumps(response, indent=2)}")
            
            # 4. List processes
            print("\n4. Listing processes...")
            response = send_json_rpc(sock, "tools/call", {
                "name": "list",
                "arguments": {}
            }, msg_id=4)
            print(f"Response: {json.dumps(response, indent=2)}")
            
            # 5. Query system overview
            print("\n5. Querying system overview...")
            response = send_json_rpc(sock, "tools/call", {
                "name": "query",
                "arguments": {
                    "type": "system_overview"
                }
            }, msg_id=5)
            print(f"Response: {json.dumps(response, indent=2)}")
            
            print("\nMCP connection test completed successfully!")
            
    except ConnectionRefusedError:
        print(f"ERROR: Could not connect to MCP server at {HOST}:{PORT}")
        print("Make sure APM is running with MCP enabled:")
        print("  APM_MCP_ENABLED=1 cargo run -- start")
        sys.exit(1)
    except Exception as e:
        print(f"ERROR: {e}")
        sys.exit(1)

if __name__ == "__main__":
    test_mcp_connection()